### PR TITLE
Use abstract byte code interface to remove unsafe/pointer code 

### PR DESCRIPTION
**Pull request by tgiphil**

### CODE REVIEW

**Comment by justinstenning** [comment]
Thanks Phil, I need to do some performance tests first on the change from unsafe and use of assemblycodefrom memory.

**Comment by tgiphil** [comment]
Great! Let me now it turns out. I could change the interface to an abstract object to cut down a few more cycles. 

I will be using this in another project that can't provide a byte array or pointer to the disassembler.

Btw. Awesome project!

**Comment by tgiphil** [comment]
I added a unit test that uses IntPtr. On my Intel Core i7-3770k @ 3.5 workstation takes 512ms to 495ms respectively for memory or byte array for 600,000 instructions covering 4,200,000 bytes.

**Comment by justinstenning** [comment]
I'm unable to check it right now but my gut is telling me there will be some significant reductions in performance, I think that perhaps it will be worth providing an IAssemblyCode class that supports the faster approach for those that can use it.

I'll try to test it out asap, however I am fairly busy so might not be for a week.

Thx glad you like it! I appreciate you taking the time to try and improve it!

**Comment by tgiphil** [comment]
I used the DisassembleLargeBuffer unit test as the benchmark and modified it to run an additional 10x iterations. I found no noticeable performance difference with the patch. 

If possible, could you push out a new **pre-release** NuGet package with this patch? That way we can use the new functionality in another project and still use NuGet package system.

File: SharpDisasm/Helpers/AssemblyCodeMemory.cs [hunk 1]
```diff
@@ -40,46 +40,44 @@
...
-    }
+		/// <summary>
+		/// Gets or sets the <see cref="System.Byte"/> at the specified index.
+		/// </summary>
+		/// <value>
+		/// The <see cref="System.Byte"/>.
+		/// </value>
+		/// <param name="index">The index.</param>
+		/// <returns></returns>
+		byte IAssemblyCode.this[int index] { get { return Marshal.ReadByte(pointer, index); }  }
```
Comment (justinstenning) [inline SharpDisasm/Helpers/AssemblyCodeMemory.cs]: I think that this should implement a range check here, i.e. if index >= Length throw new IndexOutOfRangeException();

**Comment by justinstenning** [comment]
Thanks for the effort you have put in Phil. Performance tests appear to match up pretty good.

Couple of changes needed before I can complete the pull request if that is ok:
1. can you please add an IndexOutOfRange check to the AssemblyCodeMemory this[]
2. Can you please use the Tabs as 4 spaces setting in your IDE and resave those files
3. Please change the filename comment at start of your new files to have correct filename
4. Question: is there are reason you are not exposing the IAssemblyCode implementations as public? If there is no particular reason they might as well be exposed as public to ease use if anyone makes use of the classes outside of the disassembler.

Thanks again

**Comment by tgiphil** [comment]
Thanks for the feedback. I have incorporated it into this pull request.

Note: Regarding #4, the IAssemblyCode interface is public and the implementations are now internal as generally public interfaces should not expose internal implementation unless they should be consumed elsewhere.

**Pull request merged**